rfcs/0044-ckb-light-client/0044-ckb-light-client.md

```diff
@@ -0,0 +1,508 @@
+}
+
+table GetTransactionsProof {
+    // Refer to `GetLastStateProof.last_hash`.
+    last_hash:                  Byte32,
+
+    // Transaction hashes for the transactions which require verifying.
+    tx_hashes:                  Byte32Vec,
+}
+
+table SendTransactionsProof {
+    // Refer to `SendLastStateProof.last_header`.
+    last_header:                VerifiableHeader,
+    // Refer to `SendLastStateProof.proof`.
+    proof:                      HeaderDigestVec,
+
+    // A collection of filtered blocks, which include all requested
+    // transactions, and be verified in the proof.
+    filtered_blocks:            FilteredBlockVec,
+
+    // Transaction hashes for the blocks which were not found.
+    missing_tx_hashes:          Byte32Vec,
+}
+```
+
+### Client-Server Interaction
+
+We can divide the typical client-server interaction into 3 phases.
+
+#### Phase #1: Client synchronizes to the latest tip block.
+
+- At start, the client send `GetLastState` messages to all discovered
+  servers.
+
+- All servers will reply `SendLastState` which includes their verifiable tip
+  header and the total difficulty of the chain.
+
+- The client will send `GetLastStateProof` to all servers which have replied
+  their last state. The `GetLastStateProof` messages are different for
+  different chain states, they are generated base on the sampling strategy
+  which the light client used.
+
+  _The sampling strategy is not a part of CKB. Each light client could
+  choose their own sampling strategy._
+
+  _In the official CKB light client, a sampling strategy based on the
+  [FlyClient] sampling protocol under variable difficulty is used._
+
+- The servers which received `GetLastStateProof` should reply
+  `SendLastStateProof` to the client.
+
+  If the last block that the client known is in the server's current chain,
+  then the server should return the verifiable header of the last block that
+  the client known, the proof for the sampled blocks and their verifiable
+  headers.
+
+  Otherwise, the server should only return its current verifiable tip
+  header.
+
+- Then the client will choose the best chain from all `SendBlockSamples`
+  messages.
+
+##### How a Server Choose Blocks from Sampled Difficulties?
+
+The server performs the following steps to decide which block should be
+included in the proof:
+
+- Construct a collection of blocks which is denoted as `reorg_n_blocks`.
+
+  Check if the start block, which is sent from the client, is an ancestor
+  block of the tip block.
+
+  - If true, then left `reorg_n_blocks` to be empty.
+
+  - Otherwise:
+
+    - Find the block which is an ancestor block of the tip block and its
+    number is the same as the start block.
+
+    - Add the last $L$ blocks before the block, which is found in the
+    previous step, into `reorg_n_blocks`.
+
+- Construct a collection of blocks which is denoted as `last_n_blocks`.
+
+  - Find the first block whose total difficulty is not less than
+  $D_{boundary}$.
```

Comment:
As a convention, format the multiple letter variable as `$D_{\mathrm {boundary}}$`